File: cpp/behavioral/MediatorDemo.cpp
//sbkun
#include "noncopyable.h"

#include <iostream>
#include <map>
#include <string>
#include <memory>
#include <functional>
#include <assert.h>

using namespace std;
class ChatRoom;

class User: noncopyable {
	public:
		User(const string& name);
		~User();

		void send(const string& msg);
		void recv(const string& msg);

		void setRoom(weak_ptr<ChatRoom> room) {
			room_ = room;
		}

		const string& getName() const {
			return name_;
		}

	private:
		string name_;
		weak_ptr<ChatRoom> room_;
};

//mediator
class ChatRoom : noncopyable, public enable_shared_from_this<ChatRoom>{
	public:
		typedef map<string, weak_ptr<User>> Room;

		ChatRoom(const string& name) : name_(name) {
		}

		~ChatRoom() {
			cout<<this<<" "<<name_<<" destruct"<<endl;
		}

		void addRoom(shared_ptr<User> user) {
			room_[user->getName()] = user;
			user->setRoom(shared_from_this());
		}
	

		void send( const string& msg, const string& userName) {
			for(auto const& it: room_) {
				auto pUser = it.second.lock();
				if (pUser) {
					//自己不用收自己的消息	
					if (pUser->getName() != userName) {
						pUser->recv(msg);
					}
				}
			}
		}

		const string& getName() const {
			return name_;
		}

		void removeUser(const string& userName) {
			auto it = room_.find(userName);					
			if (it != room_.end()) {
				room_.erase(it);
				cout<<"erase "<<userName<<" "<<name_<<" remain user num "<<room_.size()<<endl;
			}
		}

	private:
		Room room_;
		string name_;
};

User::User(const string& name) : name_(name)
{
}

void User::send(const string& msg) {
	shared_ptr<ChatRoom> pRoom = room_.lock();
	if (pRoom) {
		cout<<name_<<" send "<<msg<<" to "<<pRoom->getName()<<endl;
		pRoom->send(msg, name_);
	}
}

void User::recv(const string& msg) {
	shared_ptr<ChatRoom> pRoom = room_.lock();
	if (pRoom) {
		cout<<getName()<<" recv "<<msg<<" from "<<pRoom->getName()<<endl;
	}
}

User::~User() {
	cout<<this<<" "<<name_<<" destruct"<<endl;
	shared_ptr<ChatRoom> p = room_.lock();
	if (p) {
		p->removeUser(name_);
	}
}

void test() {

	auto user1 = make_shared<User>("user1");
	auto user2 = make_shared<User>("user2");
	auto user3 = make_shared<User>("user3");
	auto user4 = make_shared<User>("user4");
	auto user5 = make_shared<User>("user5");

	auto chat1 = make_shared<ChatRoom>("chat1");
	auto chat2 = make_shared<ChatRoom>("chat2");

	chat1->addRoom(user1);
	chat1->addRoom(user2);

	chat2->addRoom(user3);
	chat2->addRoom(user4);
	chat2->addRoom(user5);

	user1->send("sbkun");

	cout<<endl;

	user3->send("sbkun");

	cout<<endl;
}

void testLongLifeChatroom() {
	auto chat1 = make_shared<ChatRoom>("chat1");
	{
		auto user1 = make_shared<User>("user1");
		auto user2 = make_shared<User>("user2");
		auto user3 = make_shared<User>("user3");

		chat1->addRoom(user1);
		chat1->addRoom(user2);
		chat1->addRoom(user3);

		user1->send("sbkun");
	}
}

void testShortLifeChatroom() {
		auto user1 = make_shared<User>("user1");
		auto user2 = make_shared<User>("user2");
		auto user3 = make_shared<User>("user3");

		{
			auto chat1 = make_shared<ChatRoom>("chat1");
			chat1->addRoom(user1);
			chat1->addRoom(user2);
			chat1->addRoom(user3);

		}
		user1->send("sbkun");
}

int main(int argc, char* argv[])
{
	//testLongLifeChatroom();
	//testShortLifeChatroom();
	test();
	return 0;
}

File: java/behavioral/MediatorDemo.java
package behavioral;

import java.util.HashMap;
import java.util.Map;

class ChatRoom {
    ChatRoom(String name) {
        this.name = name;
    }

    void addRoot(User user) {
        room.put(user.getName(), user);
        user.setRoom(this);
    }

    void send(String msg, String userName) {
        for(Map.Entry<String, User> entry: room.entrySet()) {
            if (!entry.getKey().equals(userName)) {
                User user = entry.getValue();
                user.recv(msg);
            }
        }
    }

    public String getName() {
        return name;
    }

    private String name;
    private Map<String, User> room = new HashMap<String, User>();
}

class User {
    User(String name) {
        this.name = name;
    }

    void send(String msg) {
        if (room != null) {
            System.out.println(name + " send " + msg + " to " + room.getName());
            room.send(msg, name);
        }
    }

    void recv(String msg) {
        if (room != null) {
            System.out.println(name + " recv " + msg + " from " + room.getName());
        }
    }

    public String getName() {
        return name;
    }

    public void setRoom(ChatRoom room) {
        this.room = room;
    }

    private String name;

    private ChatRoom room = null;
}

public class MediatorDemo {
    static public void main(String args[]) {
        ChatRoom chat1 = new ChatRoom("chat1");
        ChatRoom chat2 = new ChatRoom("chat2");

        User user1 = new User("user1");
        User user2 = new User("user2");
        User user3 = new User("user3");
        User user4 = new User("user4");
        User user5 = new User("user5");

        chat1.addRoot(user1);
        chat1.addRoot(user2);

        chat2.addRoot(user3);
        chat2.addRoot(user4);
        chat2.addRoot(user5);

        user1.send("sbkun");

        System.out.println("");

        user3.send("sbkun");
    }
}
